subsys/bluetooth/mesh/settings.c

```diff
@@ -90,42 +74,8 @@ SETTINGS_STATIC_HANDLER_DEFINE(bt_mesh, "bt/mesh", NULL, NULL, mesh_commit,
 			      BIT(BT_MESH_SETTINGS_CFG_PENDING)      |      \
 			      BIT(BT_MESH_SETTINGS_MOD_PENDING))
 
-void bt_mesh_settings_store_schedule(enum bt_mesh_settings_flag flag)
-{
-	uint32_t timeout_ms, remaining_ms;
-
-	atomic_set_bit(pending_flags, flag);
-
-	if (atomic_get(pending_flags) & NO_WAIT_PENDING_BITS) {
-		timeout_ms = 0;
-	} else if (atomic_test_bit(pending_flags,
-				   BT_MESH_SETTINGS_RPL_PENDING) &&
-		   (!(atomic_get(bt_mesh.flags) & GENERIC_PENDING_BITS) ||
-		    (CONFIG_BT_MESH_RPL_STORE_TIMEOUT <
-		     CONFIG_BT_MESH_STORE_TIMEOUT))) {
-		timeout_ms = CONFIG_BT_MESH_RPL_STORE_TIMEOUT * MSEC_PER_SEC;
-	} else {
-		timeout_ms = CONFIG_BT_MESH_STORE_TIMEOUT * MSEC_PER_SEC;
-	}
-
-	remaining_ms = k_ticks_to_ms_floor32(k_work_delayable_remaining_get(&pending_store));
-	BT_DBG("Waiting %u ms vs rem %u ms", timeout_ms, remaining_ms);
-
-	/* If the new deadline is sooner, override any existing
-	 * deadline; otherwise schedule without changing any existing
-	 * deadline.
-	 */
-	if (timeout_ms < remaining_ms) {
-		k_work_reschedule(&pending_store, K_MSEC(timeout_ms));
-	} else {
-		k_work_schedule(&pending_store, K_MSEC(timeout_ms));
-	}
-}
-
-static void store_pending(struct k_work *work)
+static void store_pending(void)
```

Comment:
In lieu of a `settings_store_subtree` API, perhaps it makes sense to make this a public API in mesh/main.h?
Calling `settings_save()` pulls in all modules with settings storage, which could make it impossible to determine a worst case storage time.

Reply:
These are good questions! If we only need to store critical parameters of the stack then I think `settings_save()` is a wrong way and we need to make `store_pending` public and forget about `settings_save()`. Because I think if `settings_save()` supported by the stack, it should propagate this to the models, otherwise there is no other way to handle this at the models level.

However, what would be impact of not saving the models' data if power loss detected? And what if there are some other settings that need to be stored?

Reply:
@PavelVPV, @trond-snekvik, please forget about the powerloss case, it will not be possible to sustain the power long enough to be stored. For nvs with n sectors the worst case would be to do n sector **erases** before doing one save. This method would be good for planned shutdown, but a direct call to `store_pending` might be a better solution.